### PR TITLE
sync: --fix-case flag to rename case insensitive dest - fixes #4854

### LINKED ISSUE #4854: New `--fix-case` flag to workaround case insensitive filesystem (macOS/Windows)
When syncing to a case sensitive file system, it would be nice if rclone could rename files/directories to match the case on the source.

Currently this isn't possible - rclone views `HELLO.txt` and `hello.txt` on a case insensitive file system as being the same file and skips over them.

This could be controlled with a new flag `--fix-case` which would only be supported on backends which implemented the `Move` and maybe `MoveDir` optional methods.

`--fix-case` would likely rename the source to the same name as the destination if it was different. This would mean it would make the unicode normalisation the same also.

See the forum for more info: https://forum.rclone.org/t/force-rclone-to-respect-case/20794/5

### PR DESCRIPTION
**Pull request by nielash**
#### What is the purpose of this change?

Before this change, a sync to a case insensitive dest (such as macOS / Windows) would not result in a matching filename if the source and dest had casing differences but were otherwise equal. For example, syncing `hello.txt` to `HELLO.txt` would result in the dest filename remaining `HELLO.txt`. Furthermore, `--local-case-sensitive` did not solve this, as it actually caused `HELLO.txt` to get deleted!

After this change, `HELLO.txt` is renamed to `hello.txt` to match the source, only if the `--fix-case` flag is specified. (The old behavior remains the default.)

One question I have... should this maybe be the default? Can we think of a reason that the old behavior would be preferable? 

Also note that this does not address the related issues regarding `dedupe`, so perhaps #1133 or #2430 should be reopened.

#### Was the change discussed in an issue or in the forum before?

[Many times](https://forum.rclone.org/t/force-rclone-to-respect-case/20794/5) (but not by me 😄 )

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Review by ncw — COMMENTED** [review]
Very nice feature :-)

See inline for some comments.

**Comment by nielash** [comment]
Thank you @ncw! You are right, and also I just realized an additional problem: syncing `somedir/hello.txt` to `SOMEDIR/HELLO.txt` will result in `SOMEDIR/hello.txt` (with dir names still not matching.) And I'm not sure if it's wise to move the dir while other goroutines could be working on files in the same dir. Do you have any suggestions on a clean solution? I can look into it more as well. 

In the meantime, I am downgrading this PR to "draft" until I have a chance to revise.

**Comment by nielash** [comment]
Actually I think maybe I figured it out. See if I'm on the right track with this: https://github.com/rclone/rclone/pull/7356/commits/b13626bb30c2c91ad94d1fb0d28b107f5a230b74

Also, I think I found and fixed an [existing bug](https://github.com/rclone/rclone/commit/b13626bb30c2c91ad94d1fb0d28b107f5a230b74#diff-dea86f791790da6907974fcb46363041fd49f392a674366a5bef26a1d231c621) in the process (I can break this into a separate commit if you prefer)

The tests pass everywhere but windows -- I will try to debug the windows-specific issue.

**Comment by ncw** [comment]
I added some comments!

If you separate the fix into a new commit that would be perfect. Or a new PR. You could make a failing test case which you make pass with the fix.

File: fs/config/configflags/configflags.go [hunk 1]
```diff
@@ -82,6 +82,7 @@ func AddFlags(ci *fs.ConfigInfo, flagSet *pflag.FlagSet) {
 	flags.BoolVarP(flagSet, &ci.IgnoreSize, "ignore-size", "", false, "Ignore size when skipping use mod-time or checksum", "Copy")
 	flags.BoolVarP(flagSet, &ci.IgnoreChecksum, "ignore-checksum", "", ci.IgnoreChecksum, "Skip post copy check of checksums", "Copy")
 	flags.BoolVarP(flagSet, &ci.IgnoreCaseSync, "ignore-case-sync", "", ci.IgnoreCaseSync, "Ignore case when synchronizing", "Copy")
+	flags.BoolVarP(flagSet, &ci.FixCase, "fix-case", "", ci.FixCase, "Force rename of case insensitive dest to match source", "Copy")
```
[the remark this answers is not in the document]
Comment (nielash) [reply fs/config/configflags/configflags.go]: Fixed in latest version.

File: fs/sync/sync.go [hunk 2]
```diff
@@ -378,6 +378,15 @@ func (s *syncCopyMove) pairChecker(in *pipe, out *pipe, fraction int, wg *sync.W
 					}
 				}
 			} else {
+				// Fix case for case insensitive filesystems
+				if s.ci.FixCase && src.Remote() != pair.Dst.Remote() {
```
[the remark this answers is not in the document]
Comment (nielash) [reply fs/sync/sync.go]: Fixed in latest version.

File: fs/sync/sync_test.go [hunk 3]
```diff
@@ -1999,6 +1999,33 @@ func TestSyncIgnoreCase(t *testing.T) {
...
+
+	// Only test if at least one filesystem is case insensitive
+	if !r.Fremote.Features().CaseInsensitive && !r.Flocal.Features().CaseInsensitive {
+		t.Skip("Skipping test as local or remote are case-sensitive")
+	}
+
+	ci.FixCase = true
+
+	// Create files with different filename casing
+	file1 := r.WriteFile("existing", "potato", t1)
```
[the remark this answers is not in the document]
Comment (nielash) [reply fs/sync/sync_test.go]: Fixed in latest version.

**Comment by nielash** [comment]
Fixed the windows-specific issue (for the record: it was [this](https://stackoverflow.com/questions/12132125/golang-cross-platform-path-dir/12132445#12132445).)

**Comment by nielash** [comment]
Closing as a dupe of #7410, which is now merged.

**Pull request closed without merging**